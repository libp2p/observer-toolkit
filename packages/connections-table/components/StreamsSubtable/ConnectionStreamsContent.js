import React from 'react'
import T from 'prop-types'
import styled from 'styled-components'

import { Icon, StyledButton, Tooltip } from '@libp2p-observer/sdk'

import StreamsSubtable from './StreamsSubtable'

const Container = styled.span`
  text-align: right;
  white-space: nowrap;
`

const StreamsNumber = styled.span`
  margin-right: ${({ theme }) => theme.spacing(2)};
  min-width: ${({ theme }) => theme.spacing(4)};
  display: inline-block;
`

const ExpandIcon = styled.span`
  transform: rotate(90deg);
  margin-right: ${({ theme }) => theme.spacing(-1)};
  [data-tooltip='open'] > button > & {
    transform: rotate(180deg);
  }
  ${({ theme }) => theme.transition({ property: 'transform' })}
`

const ExpandButton = styled(StyledButton)`
  ${({ theme }) => theme.transition()}
  [data-tooltip="open"] > & {
    background: ${({ theme }) => theme.color('background', 1)};
  }
`

function ConnectionStreamsContent({
  value,
  streamsCount = value,
  connection,
  hidePrevious,
}) {
  const streamsButtonText = streamsCount ? `View streams` : 'No streams'

  return (
    <Container>
      <StreamsNumber>{streamsCount}</StreamsNumber>
      <Tooltip
        side="bottom"
        fixOn="no-hover"
        toleranceY={null}
        toleranceX={-32}
        hidePrevious={hidePrevious}
        content={<StreamsSubtable connection={connection} />}
      >
        <ExpandButton disabled={!streamsCount}>
          {streamsButtonText}
          {streamsCount && (
            <Icon type="expand" override={{ Container: ExpandIcon }} />
          )}
        </ExpandButton>
      </Tooltip>
    </Container>
  )
}

ConnectionStreamsContent.propTypes = {
  value: T.number,
  connection: T.object.isRequired,
  streamsCount: T.number,
  hidePrevious: T.func,
}

export default ConnectionStreamsContent
