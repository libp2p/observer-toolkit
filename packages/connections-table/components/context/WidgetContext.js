import React, { useContext, useMemo } from 'react'
import T from 'prop-types'

import {
  getConnections,
  getConnectionTraffic,
  getConnectionAge,
  getConnectionTimeClosed,
  getStreams,
  statusNames,
} from '@nearform/observer-data'
import {
  formatDuration,
  formatDataSize,
  getListFilter,
  getRangeFilter,
  useCalculation,
  FilterProvider,
  DataContext,
  TimeContext,
} from '@nearform/observer-sdk'

import { MetadataProvider } from './MetadataProvider'

const statusNamesList = Object.values(statusNames)

function getMaxValues(states) {
  const maxValues = states.reduce(
    (maxValues, state) =>
      getConnections(state).reduce((timeMax, connection) => {
        const { maxTraffic, maxAge } = timeMax
        const dataIn = getConnectionTraffic(connection, 'in', 'bytes')
        const dataOut = getConnectionTraffic(connection, 'out', 'bytes')
        const age = getConnectionAge(connection, state)
        return {
          maxTraffic: Math.max(maxTraffic, dataIn, dataOut),
          maxAge: Math.max(maxAge, age),
        }
      }, maxValues),
    {
      maxTraffic: 0,
      maxAge: 0,
    }
  )
  return maxValues
}

function WidgetContext({ children }) {
  const states = useContext(DataContext)
  const currentState = useContext(TimeContext)

  // If performance becomes an issue on live-streaming data, use
  // useReducer and compare appended data only instead of whole dataset
  const metadata = useMemo(() => getMaxValues(states), [states])

  const maxStreams = useCalculation(
    ({ currentState }) =>
      getConnections(currentState).reduce(
        (max, conn) => Math.max(max, getStreams(conn).length),
        0
      ),
    ['currentState']
  )

  const filterDefs = [
    getListFilter({
      name: 'Filter by status',
      mapFilter: conn => statusNamesList[conn.getStatus()],
      valueNames: statusNamesList,
    }),
    getRangeFilter({
      name: 'Filter number of streams',
      mapFilter: conn => getStreams(conn).length,
      min: 0,
      max: maxStreams,
    }),
    getRangeFilter({
      name: 'Filter total bytes in',
      mapFilter: conn => getConnectionTraffic(conn, 'in', 'bytes'),
      min: 0,
      max: metadata.maxTraffic,
      format: formatDataSize,
    }),
    getRangeFilter({
      name: 'Filter total bytes out',
      mapFilter: conn => getConnectionTraffic(conn, 'out', 'bytes'),
      min: 0,
      max: metadata.maxTraffic,
      format: formatDataSize,
    }),
    getRangeFilter({
      name: 'Filter miliseconds open',
      mapFilter: conn => getConnectionAge(conn, currentState),
      min: 0,
      max: metadata.maxAge,
      format: formatDuration,
    }),
    getRangeFilter({
      name: 'Filter miliseconds closed',
      mapFilter: conn => getConnectionTimeClosed(conn, currentState),
      min: 0,
      max: metadata.maxAge,
      format: formatDuration,
    }),
  ]

  return (
    <MetadataProvider metadata={metadata}>
      <FilterProvider filterDefs={filterDefs}>{children}</FilterProvider>
    </MetadataProvider>
  )
}

WidgetContext.propTypes = {
  children: T.node,
}

export default WidgetContext
