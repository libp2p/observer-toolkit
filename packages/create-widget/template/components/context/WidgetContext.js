import React, { useContext } from 'react'
import T from 'prop-types'

import { getStateRangeTimes } from '@libp2p/observer-data'
import {
  getRangeFilter,
  FilterProvider,
  DataContext,
} from '@libp2p/observer-sdk'

// WidgetContext is wrapped arpund every widget and its outer container,
// allowing contexts to share with any component within the widget.
function WidgetContext({ children }) {
  const states = useContext(DataContext)
  const { start, end } = getStateRangeTimes(states)

  // Filter definitions provided to FilterProvider in the WidgetContext
  // will be attached to the widget container's filter tray
  const filterDefs = [
    getRangeFilter({
      name: 'Filter events by event time',
      mapFilter: event => event.getTs(),
      min: start,
      max: end,
      numberFieldType: 'time',
    }),
  ]

  return <FilterProvider filterDefs={filterDefs}>{children}</FilterProvider>
}

WidgetContext.propTypes = {
  children: T.node,
}

export default WidgetContext
