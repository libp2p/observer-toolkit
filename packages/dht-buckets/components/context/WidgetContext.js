import React from 'react'
import T from 'prop-types'

import {
  dhtQueryDirectionNames,
  dhtQueryResultNames,
} from '@libp2p-observer/data'
import {
  getListFilter,
  getRangeFilter,
  FilterProvider,
} from '@libp2p-observer/sdk'

import { DhtQueryProvider } from './DhtQueryProvider'
import { PeerSlotsProvider } from './PeerSlotsProvider'

const directionNamesList = Object.values(dhtQueryDirectionNames)
const resultNamesList = Object.values(dhtQueryResultNames)

function WidgetContext({ children }) {
  const filterDefs = [
    getListFilter({
      name: 'Filter query direction',
      mapFilter: query => directionNamesList[query.getDirection()],
      valueNames: directionNamesList,
    }),
    getListFilter({
      name: 'Filter query result',
      mapFilter: query => resultNamesList[query.getResult()],
      valueNames: resultNamesList,
    }),
    getRangeFilter({
      name: 'Filter query duration',
      mapFilter: query => query.duration,
      min: 0,
      max: 1000, // TODO: See how real libp2p data exposes query limit
    }),
  ]

  return (
    <FilterProvider filterDefs={filterDefs}>
      <PeerSlotsProvider>
        <DhtQueryProvider>{children}</DhtQueryProvider>
      </PeerSlotsProvider>
    </FilterProvider>
  )
}

WidgetContext.propTypes = {
  children: T.node,
}

export default WidgetContext
