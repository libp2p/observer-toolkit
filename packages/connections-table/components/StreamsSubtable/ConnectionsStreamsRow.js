import React, { useContext } from 'react'
import T from 'prop-types'
import styled from 'styled-components'

import { Icon, TableCell, TableRow, TimeContext } from '@libp2p-observer/sdk'
import { getConnections } from '@libp2p-observer/data'
import StreamsSubtable from './StreamsSubtable'

const PushApart = styled.div`
  white-space: nowrap;
  display: flex;
  flex-direction: row;
  margin: ${({ theme }) => theme.spacing()};
  align-items: center;
  justify-content: space-between;
`

const ExpandedRow = styled(TableRow)`
  background: none;
  border: solid 8px
    ${({ theme, highlighted }) =>
      highlighted
        ? theme.color('background', 2)
        : theme.color('background', 1)};
`

const ConnectionCell = styled(TableCell)`
  vertical-align: top;
  border: inherit;
`

const StreamsCell = styled(TableCell)`
  border: inherit;
  position: relative;
  vertical-align: top;
  padding: 0;
`

const Heading = styled.h3`
  ${({ theme }) => theme.text('heading', 'small')};
  color: ${({ theme }) => theme.color('highlight')};
`

const CloseIcon = styled.span``

function ConnectionsStreamsRow({
  peerId,
  streamsCount,
  rowContentProps,
  columnDefs,
  closeRow,
  onMouseEnter,
  onMouseLeave,
  highlighted,
}) {
  const timepoint = useContext(TimeContext)
  const connections = getConnections(timepoint)
  const connection = connections.find(conn => conn.getPeerId() === peerId)

  return (
    <ExpandedRow
      highlighted={highlighted}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      <ConnectionCell colSpan={2}>
        {columnDefs.map(({ renderContent, name, header }, cellIndex) => {
          const capitalisedInitial = header[0].toUpperCase() + header.slice(1)
          const labelId = `${name}_${cellIndex}_${peerId}`
          return (
            <PushApart key={name}>
              <label htmlFor={labelId}>{capitalisedInitial}:</label>
              <span id={labelId}>
                {renderContent(rowContentProps[cellIndex])}
              </span>
            </PushApart>
          )
        })}
      </ConnectionCell>
      <StreamsCell colSpan={columnDefs.length - 1}>
        <PushApart>
          <Heading>Connection has {streamsCount} streams:</Heading>
          <Icon
            aria-label="Close"
            type="cancel"
            active
            size="2.5em"
            onClick={closeRow}
            override={{ Container: CloseIcon }}
          />
        </PushApart>
        <StreamsSubtable connection={connection} />
      </StreamsCell>
    </ExpandedRow>
  )
}

ConnectionsStreamsRow.propTypes = {
  peerId: T.string.isRequired,
  streamsCount: T.number.isRequired,
  rowContentProps: T.array.isRequired,
  columnDefs: T.array.isRequired,
  closeRow: T.func.isRequired,
  onMouseEnter: T.func,
  onMouseLeave: T.func,
  highlighted: T.bool,
}

export default ConnectionsStreamsRow
