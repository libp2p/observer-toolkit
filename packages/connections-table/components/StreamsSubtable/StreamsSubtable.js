import React, { useContext } from 'react'
import T from 'prop-types'

import { getStreams } from '@libp2p-observer/data'
import { DataTable, useTabularData } from '@libp2p-observer/sdk'

import { MetadataContext } from '../context/MetadataProvider'
import streamsColumnDefs from '../../definitions/streamsColumns'

function StreamsSubTable({ connection }) {
  const metadata = useContext(MetadataContext)

  const {
    columnDefs,
    contentProps,
    sortColumn,
    setSortColumn,
    sortDirection,
    setSortDirection,
  } = useTabularData({
    columns: streamsColumnDefs,
    data: getStreams(connection),
    defaultSort: 'stream-status',
    metadata,
  })

  return (
    <DataTable
      contentProps={contentProps}
      columnDefs={columnDefs}
      sortColumn={sortColumn}
      setSortColumn={setSortColumn}
      sortDirection={sortDirection}
      setSortDirection={setSortDirection}
    />
  )
}

StreamsSubTable.propTypes = {
  connection: T.object,
}

export default StreamsSubTable
