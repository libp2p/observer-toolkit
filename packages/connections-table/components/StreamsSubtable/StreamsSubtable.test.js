import React from 'react'
import { loadSample, renderWithData, within } from '@nearform/observer-testing'
import {
  getConnections,
  getStreams,
  getLatestTimepoint,
} from '@nearform/observer-data'
import StreamsSubtable from './StreamsSubtable'
import WidgetContext from '../context/WidgetContext'

describe('StreamsSubtable', () => {
  const {
    data: { states },
  } = loadSample()
  const timepoint = getLatestTimepoint(states)

  const [connection] = getConnections(timepoint)
  const StreamsSubtableInContext = () => (
    <WidgetContext>
      <StreamsSubtable connection={connection} />
    </WidgetContext>
  )

  it('has a row for each stream in the connection, paginated', () => {
    const { getAllByTableRow, getByRole } = renderWithData(
      <StreamsSubtableInContext />
    )
    const streamsCount = getStreams(connection).length
    const rows = getAllByTableRow()
    const perPage = getByRole(/(listbox|combobox)/, {
      selector: '[name="PerPage"]',
    })
    const [perPageSelectedOption] = [
      ...within(perPage).getAllByRole('option'),
    ].filter(elem => elem.value === perPage.value)
    const rowsPerPage = Number(perPageSelectedOption.textContent)

    expect(rows).toHaveLength(Math.min(streamsCount, rowsPerPage))
  })
})
