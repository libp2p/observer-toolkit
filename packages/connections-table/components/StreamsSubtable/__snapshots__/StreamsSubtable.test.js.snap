// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreamsSubtable renders connection's streams as expected 1`] = `
<DocumentFragment>
  .c0 {
  width: inherit;
  height: inherit;
  display: inherit;
  -webkit-flex-direction: inherit;
  -ms-flex-direction: inherit;
  flex-direction: inherit;
}

.c5 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  vertical-align: middle;
  width: 2em;
  height: 2em;
}

.c7 {
  display: inline-block;
}

.c14 {
  position: relative;
  display: inline-block;
  cursor: pointer;
}

.c11 {
  display: inline-block;
  padding: 0px 4px;
  background: rgba(163,26,75,0.15);
  color: rgba(163,26,75,1);
  white-space: nowrap;
  font-size: 8pt;
  font-weight: 600;
}

.c12 {
  line-height: 2em;
  display: inline-block;
  padding: 4px;
  vertical-align: middle;
}

.c1 {
  border-collapse: collapse;
  width: 100%;
}

.c9 {
  border-bottom: 1px solid rgba(241,242,240,0.5);
}

.c2 {
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 9pt;
  line-height: 1.2em;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 24px;
  padding-right: 16px;
  text-align: left;
  padding-right: 0;
  font-family: plex-sans;
  font-weight: 600;
  background-color: rgba(32,155,198,0.2);
  color: rgba(93,94,96,0.8);
  text-transform: uppercase;
  white-space: nowrap;
}

.c8 {
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 9pt;
  line-height: 1.2em;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 24px;
  padding-right: 16px;
  text-align: left;
  padding-right: 0;
  font-family: plex-sans;
  font-weight: 600;
  background-color: rgba(32,155,198,0.2);
  color: rgba(93,94,96,0.8);
  text-transform: uppercase;
  white-space: nowrap;
  text-align: right;
}

.c10 {
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 9pt;
  line-height: 1.2em;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 24px;
  padding-right: 16px;
  text-align: left;
  padding-right: 32px;
  font-weight: 400;
  font-family: plex-sans;
  color: rgb(93,94,96);
}

.c13 {
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 9pt;
  line-height: 1.2em;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 24px;
  padding-right: 16px;
  text-align: left;
  padding-right: 32px;
  font-weight: 400;
  font-family: plex-sans;
  color: rgb(93,94,96);
  text-align: right;
}

.c4 {
  height: 2em;
  width: 2em;
  display: inline-block;
  vertical-align: middle;
  margin-left: 4px;
  -webkit-transition: -webkit-transform 0.4s ease-in-out;
  -webkit-transition: transform 0.4s ease-in-out;
  transition: transform 0.4s ease-in-out;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c6 {
  height: 2em;
  width: 2em;
  display: inline-block;
  vertical-align: middle;
  margin-left: 4px;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c3 {
  cursor: pointer;
}

.c16 {
  font-family: 'plex-mono';
  white-space: nowrap;
}

.c17 {
  margin-right: -8px;
  padding-left: 8px;
  font-family: 'plex-sans';
  font-weight: 300;
  font-size: 90%;
  width: 32px;
  display: inline-block;
  text-align: left;
}

.c18 {
  border-radius: 50%;
  height: 24px;
  width: 24px;
  display: inline-block;
  vertical-align: middle;
  margin-left: 8px;
  margin-right: -16px;
}

.c19 {
  border-radius: 50%;
  border: 2px solid rgba(255,255,255,0.8);
  background: rgb(252,189,32);
}

.c20 {
  border-radius: 50%;
  border: 2px solid rgba(255,255,255,0.8);
  background: rgb(112,208,224);
}

.c21 {
  border-radius: 50%;
  border: 2px solid rgba(255,255,255,0.8);
  background: currentColor;
}

.c15 {
  white-space: nowrap;
}

<div
    class="c0"
  >
    <table
      class="c1"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class="c2"
          >
            <div
              class="c3"
            >
              status
              <span
                class="c4"
              >
                <span
                  class="c5"
                  role="img"
                >
                  <svg
                    viewBox="0 0 32 32"
                  >
                    <path
                      d="M18.933 24.716a.85.85 0 00.601-1.451l-7.264-7.263 7.264-7.267a.85.85 0 00-1.201-1.202l-7.866 7.868a.852.852 0 00.001 1.203l7.865 7.864a.846.846 0 00.6.248z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </th>
          <th
            class="c2"
          >
            <div
              class="c3"
            >
              protocol
              <span
                class="c6"
              >
                <span
                  class="c5"
                  role="img"
                >
                  <span
                    class="c7"
                  />
                </span>
              </span>
            </div>
          </th>
          <th
            class="c8"
          >
            <div
              class="c3"
            >
              Data in
              <span
                class="c6"
              >
                <span
                  class="c5"
                  role="img"
                >
                  <span
                    class="c7"
                  />
                </span>
              </span>
            </div>
          </th>
          <th
            class="c8"
          >
            <div
              class="c3"
            >
              Data out
              <span
                class="c6"
              >
                <span
                  class="c5"
                  role="img"
                >
                  <span
                    class="c7"
                  />
                </span>
              </span>
            </div>
          </th>
          <th
            class="c2"
          >
            <div
              class="c3"
            >
              Time open
              <span
                class="c6"
              >
                <span
                  class="c5"
                  role="img"
                >
                  <span
                    class="c7"
                  />
                </span>
              </span>
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      >
        <tr
          class="c9"
        >
          <td
            class="c10"
          >
            <span
              class="c11"
            >
              <span
                class="c5"
                role="img"
              >
                <svg
                  viewBox="0 0 32 32"
                >
                  <path
                    d="M16 7.387a8.612 8.612 0 018.612 8.615c0 4.752-3.854 8.606-8.612 8.606-4.756 0-8.61-3.854-8.61-8.606A8.61 8.61 0 0116 7.387m0-1.188c-5.412 0-9.8 4.389-9.8 9.803 0 5.41 4.388 9.799 9.8 9.799 5.413 0 9.801-4.389 9.801-9.799 0-5.414-4.389-9.803-9.801-9.803z"
                    fill="currentColor"
                  />
                  <path
                    d="M10.654 15.287H21.27v1.425H10.654z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="c12"
              >
                CLOSED
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            other-protocol
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  835
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c19"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  835
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c20"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  1576150655
                  <span
                    class="c17"
                  >
                    s
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c21"
                  />
                </div>
              </span>
            </span>
          </td>
        </tr>
        <tr
          class="c9"
        >
          <td
            class="c10"
          >
            <span
              class="c11"
            >
              <span
                class="c5"
                role="img"
              >
                <svg
                  viewBox="0 0 32 32"
                >
                  <path
                    d="M16 7.387a8.612 8.612 0 018.612 8.615c0 4.752-3.854 8.606-8.612 8.606-4.756 0-8.61-3.854-8.61-8.606A8.61 8.61 0 0116 7.387m0-1.188c-5.412 0-9.8 4.389-9.8 9.803 0 5.41 4.388 9.799 9.8 9.799 5.413 0 9.801-4.389 9.801-9.799 0-5.414-4.389-9.803-9.801-9.803z"
                    fill="currentColor"
                  />
                  <path
                    d="M10.654 15.287H21.27v1.425H10.654z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="c12"
              >
                CLOSED
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            other-protocol
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  922
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c19"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  922
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c20"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  1576150655
                  <span
                    class="c17"
                  >
                    s
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c21"
                  />
                </div>
              </span>
            </span>
          </td>
        </tr>
        <tr
          class="c9"
        >
          <td
            class="c10"
          >
            <span
              class="c11"
            >
              <span
                class="c5"
                role="img"
              >
                <svg
                  viewBox="0 0 32 32"
                >
                  <path
                    d="M16 7.387a8.612 8.612 0 018.612 8.615c0 4.752-3.854 8.606-8.612 8.606-4.756 0-8.61-3.854-8.61-8.606A8.61 8.61 0 0116 7.387m0-1.188c-5.412 0-9.8 4.389-9.8 9.803 0 5.41 4.388 9.799 9.8 9.799 5.413 0 9.801-4.389 9.801-9.799 0-5.414-4.389-9.803-9.801-9.803z"
                    fill="currentColor"
                  />
                  <path
                    d="M10.654 15.287H21.27v1.425H10.654z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="c12"
              >
                CLOSED
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            other-protocol
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  748
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c19"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  748
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c20"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  1576150656
                  <span
                    class="c17"
                  >
                    s
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c21"
                  />
                </div>
              </span>
            </span>
          </td>
        </tr>
        <tr
          class="c9"
        >
          <td
            class="c10"
          >
            <span
              class="c11"
            >
              <span
                class="c5"
                role="img"
              >
                <svg
                  viewBox="0 0 32 32"
                >
                  <path
                    d="M16 7.387a8.612 8.612 0 018.612 8.615c0 4.752-3.854 8.606-8.612 8.606-4.756 0-8.61-3.854-8.61-8.606A8.61 8.61 0 0116 7.387m0-1.188c-5.412 0-9.8 4.389-9.8 9.803 0 5.41 4.388 9.799 9.8 9.799 5.413 0 9.801-4.389 9.801-9.799 0-5.414-4.389-9.803-9.801-9.803z"
                    fill="currentColor"
                  />
                  <path
                    d="M10.654 15.287H21.27v1.425H10.654z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="c12"
              >
                CLOSED
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            some-protocol
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  931
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c19"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  931
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c20"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  1576150655
                  <span
                    class="c17"
                  >
                    s
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c21"
                  />
                </div>
              </span>
            </span>
          </td>
        </tr>
        <tr
          class="c9"
        >
          <td
            class="c10"
          >
            <span
              class="c11"
            >
              <span
                class="c5"
                role="img"
              >
                <svg
                  viewBox="0 0 32 32"
                >
                  <path
                    d="M16 7.387a8.612 8.612 0 018.612 8.615c0 4.752-3.854 8.606-8.612 8.606-4.756 0-8.61-3.854-8.61-8.606A8.61 8.61 0 0116 7.387m0-1.188c-5.412 0-9.8 4.389-9.8 9.803 0 5.41 4.388 9.799 9.8 9.799 5.413 0 9.801-4.389 9.801-9.799 0-5.414-4.389-9.803-9.801-9.803z"
                    fill="currentColor"
                  />
                  <path
                    d="M10.654 15.287H21.27v1.425H10.654z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="c12"
              >
                CLOSED
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            other-protocol
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  63.2
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c19"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  63.2
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c20"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  1576150655
                  <span
                    class="c17"
                  >
                    s
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c21"
                  />
                </div>
              </span>
            </span>
          </td>
        </tr>
        <tr
          class="c9"
        >
          <td
            class="c10"
          >
            <span
              class="c11"
            >
              <span
                class="c5"
                role="img"
              >
                <svg
                  viewBox="0 0 32 32"
                >
                  <path
                    d="M16 7.387a8.612 8.612 0 018.612 8.615c0 4.752-3.854 8.606-8.612 8.606-4.756 0-8.61-3.854-8.61-8.606A8.61 8.61 0 0116 7.387m0-1.188c-5.412 0-9.8 4.389-9.8 9.803 0 5.41 4.388 9.799 9.8 9.799 5.413 0 9.801-4.389 9.801-9.799 0-5.414-4.389-9.803-9.801-9.803z"
                    fill="currentColor"
                  />
                  <path
                    d="M10.654 15.287H21.27v1.425H10.654z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="c12"
              >
                CLOSED
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            some-protocol
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  500
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c19"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  500
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c20"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  1576150655
                  <span
                    class="c17"
                  >
                    s
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c21"
                  />
                </div>
              </span>
            </span>
          </td>
        </tr>
        <tr
          class="c9"
        >
          <td
            class="c10"
          >
            <span
              class="c11"
            >
              <span
                class="c5"
                role="img"
              >
                <svg
                  viewBox="0 0 32 32"
                >
                  <path
                    d="M16 7.387a8.612 8.612 0 018.612 8.615c0 4.752-3.854 8.606-8.612 8.606-4.756 0-8.61-3.854-8.61-8.606A8.61 8.61 0 0116 7.387m0-1.188c-5.412 0-9.8 4.389-9.8 9.803 0 5.41 4.388 9.799 9.8 9.799 5.413 0 9.801-4.389 9.801-9.799 0-5.414-4.389-9.803-9.801-9.803z"
                    fill="currentColor"
                  />
                  <path
                    d="M10.654 15.287H21.27v1.425H10.654z"
                    fill="currentColor"
                  />
                </svg>
              </span>
              <span
                class="c12"
              >
                CLOSED
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            other-protocol
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  49.2
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c19"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c13"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  49.2
                  <span
                    class="c17"
                  >
                    mb
                  </span>
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c20"
                  />
                </div>
              </span>
            </span>
          </td>
          <td
            class="c10"
          >
            <span
              class="c14"
            >
              <span
                class="c15"
              >
                <span
                  class="c16"
                >
                  0
                  <span
                    class="c17"
                  />
                </span>
                <div
                  class="c18"
                  size="24"
                >
                  <div
                    class="c21"
                  />
                </div>
              </span>
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
